app/assets/javascripts/discourse/mixins/pan-events.js.es6

```diff
@@ -0,0 +1,89 @@
+import {
+  on,
+  default as computed
+} from "ember-addons/ember-computed-decorators";
+
+export default Ember.Mixin.create({
+
+  //velocity is pixels per ms
+
+  _panState: null,
+
+  didInsertElement() {
+    this.$().on("pointerdown", (e) => this._panStart(e))
+      .on("pointermove", (e) => this._panMove(e))
+      .on("pointerup", (e) => this._panMove(e))
+      .on("pointercancel", (e) => this._panMove(e));
+  },
+
+  willDestroyElement() {
+    this.$().off("pointerdown")
+      .off("pointerup")
+      .off("pointermove")
+      .off("pointercancel");
+  },
+
+  _calculateNewPanState(oldState, e) {
+    if(e.type == "pointerup" || e.type == "pointercancel") {
+      return oldState;
+    }
+    const newTimestamp = new Date().getTime();
+    const timeDiffSeconds = (newTimestamp - oldState.timestamp);
+    //calculate delta x, y, distance from START location
+    const deltaX = Math.round(e.clientX) - oldState.startLocation.x;
+    const deltaY = Math.round(e.clientY) - oldState.startLocation.y;
+    const distance = Math.round(Math.sqrt(Math.pow(deltaX, 2) + Math.pow(deltaY, 2)));
+
+    //calculate velocity from previous event center location
+    const eventDeltaX = e.clientX - oldState.center.x;
+    const eventDeltaY = e.clientY - oldState.center.y;
+    const velocityX = eventDeltaX / timeDiffSeconds;
+    const velocityY = eventDeltaY / timeDiffSeconds;
+    const deltaDistance = Math.sqrt(Math.pow(eventDeltaX, 2) + Math.pow(eventDeltaY, 2));
+    const velocity = deltaDistance / timeDiffSeconds;
+
+    return {
```

Comment:
In es6 you don’t need to repeat the value in objects if key has the text than value variable:

```
return {
   startLocation: oldState.startLocation,
   center: {x: Math.round(e.clientX), y: Math.round(e.clientY)},
   velocity,
   velocityX,
   velocityY,
   deltaX,
   deltaY,
   distance,
   start: false,
   timestamp: newTimestamp
}
```